src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -143,15 +150,20 @@ public TextDocumentState UpdateText(SourceText newText, PreservationMode mode)
         return UpdateText(newTextAndVersion, mode);
     }
 
-    public TextDocumentState UpdateText(TextLoader loader, PreservationMode mode)
+    public TextDocumentState UpdateText(TextLoader? loader, PreservationMode mode)
     {
         // don't blow up on non-text documents.
-        var newTextSource = CreateTextFromLoader(loader, mode, this.SolutionServices);
+        var newTextSource = CreateTextAndVersionSource(SolutionServices, loader, FilePath, LoadTextOptions, mode);
 
         return UpdateText(newTextSource, mode, incremental: false);
     }
 
-    private static ITextAndVersionSource CreateTextFromLoader(TextLoader loader, PreservationMode mode, SolutionServices solutionServices)
+    protected static ITextAndVersionSource CreateTextAndVersionSource(SolutionServices solutionServices, TextLoader? loader, string? filePath, LoadTextOptions loadTextOptions, PreservationMode mode = PreservationMode.PreserveValue)
+        => loader != null
+            ? CreateTextFromLoader(solutionServices, loader, mode)
+            : CreateStrongText(TextAndVersion.Create(SourceText.From(string.Empty, encoding: null, loadTextOptions.ChecksumAlgorithm), VersionStamp.Default, filePath));
```

Comment:
not having an encoding always worries me as a timebomb for something breaking down the line.

Reply:
No change in logic here, just extracted to a method.

Reply:
Actually, I already did that in previous PR :), so this just changed location of the method. WIll move back.

Reply:
gotcha.  i still hate it :)